scipy/linalg/_basic.py

```diff
@@ -261,6 +309,45 @@ def solve(a, b, lower=False, overwrite_a=False,
     return x
 
 
+def _lange_diagonal(_, a):
+    # Equivalent of dlange for diagonal matrix, assuming
+    # norm is either 'I' or '1' (really just not the Frobenius norm)
+    return np.abs(np.diag(a)).max()
+
+
+def _lange_tridiagonal(norm, a):
+    # Equivalent of dlange for tridiagonal matrix, assuming
+    # norm is either 'I' or '1'
+    if norm == 'I':
+        a = a.T
+    dl, d, du = np.diag(a, -1), np.diag(a, 0), np.diag(a, 1)
+    a = abs(dl[0]) + abs(d[0])
+    b = (np.abs(dl[1:]) + np.abs(d[1:-1]) + np.abs(du[:-1])).max()
+    c = abs(du[-1]) + abs(d[-1])
+    return max(a, b, c)
+
+
+def _bandwidth(a):
```

Comment:
K. At the beginning of `solve` I will coerce to one of those types. (Why is that not already done? Isn't there at least a private function for this?)

Reply:
In the previous state of this PR there was a casting table and I was doing it with a casting table etc. but probably now overwritten.

Reply:
K. Just seems fundamental if all of `linalg` is based on LAPACK which accepts only the four flavors. Seems there is no harm in taking care of the conversion early if LAPACK will be used rather than run the risk of letting multiple calls to wrappers do the same conversions repeatedly.

Reply:
Yes it is a major pain to use BLAS and LAPACK to massage things until functions accept the input.

Reply:
Hopefully the approach I took looks OK. It seems to work for typical dtypes.